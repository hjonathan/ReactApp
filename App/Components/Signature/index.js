import React, {PureComponent} from "react";
import {View, Button} from "react-native";
import SignatureCapture from "react-native-signature-capture";
import Orientation from "react-native-orientation";
import PropTypes from "prop-types";
import I18n from "../../I18n";
import Styles from "./styles";
import * as Res from "../../Assets/resources";

/**
 * Class Signature
 * Example:
 * import Signature from '../Components/Signature';
 * render() {
 *      return(
 *          <Signature
 *              // This a function where the signature object returns
 *              // objectSignature.encoded - for the base64 encoded png
 *              // objectSignature.pathName - for the file path name
 *              sendSignature={(objectSignature) => {
 *                  console.log(objectSignature);
 *              }}
 *          />
 *      ),
 * }
 */
class Signature extends PureComponent {
    /**
     * Constructor Signature
     * @param props
     */
    constructor (props) {
        super(props);
        this.state = {
            saveDisable: true,
            orientation: null
        };
        this.signature = null;
        this.onSaveEvent = this.onSaveEvent.bind(this);
        this.onDragEvent = this.onDragEvent.bind(this);
        this.onResetButton = this.onResetButton.bind(this);
    }

    /**
     * Component will be mounted.
     */
    componentWillMount () {
        const initial = Orientation.getInitialOrientation();
        this.setState({orientation: initial});
    }

    /**
     * Reset the cases state when the component was removed
     */
    componentWillUnmount () {
        Orientation.unlockAllOrientations();
    }

    /**
     * Cancel button in modal
     */
    onCancelButton () {
        this.props.changeSignature(false);
        if (this.state.orientation === "LANDSCAPE") {
            Orientation.lockToLandscape();
        } else {
            Orientation.lockToPortrait();
        }
    }

    /**
     * Reset canvas signature
     */
    onResetButton () {
        this.signature.resetImage();
        this.setState({saveDisable: true});
    }

    /**
     * Save the signature
     */
    onSaveButton () {
        this.signature.saveImage();
        setTimeout(() => {
            this.props.changeSignature(false);
            if (this.state.orientation === "LANDSCAPE") {
                Orientation.lockToLandscape();
            } else {
                Orientation.lockToPortrait();
            }
        }, 1000);
    }

    /**
     * Save event
     * @param result
     */
    onSaveEvent (result) {
        this.props.sendSignature(result);
    }

    /**
     * Drag event
     */
    onDragEvent () {
        this.setState({saveDisable: false});
    }

    /**
     * Sets ref to the signature picture
     * @param ref
     * @returns {*}
     */
    setSignatureRef = ref => (this.signature = ref);

    /**
     * Renders Signature
     */
    render () {
        return (
            <View style={Styles.modalContainer}>
                <View style={Styles.modalOptions}>
                    <View style={Styles.buttonStyle}>
                        <Button
                            color={Res.colors.gray}
                            onPress={() => {
                                this.onCancelButton();
                            }}
                            title={I18n.t("signature_cancel")}
                        />
                    </View>
                    <View style={Styles.buttonStyle}>
                        <Button
                            color={Res.colors.gray}
                            onPress={() => {
                                this.onResetButton();
                            }}
                            title={I18n.t("signature_clear")}
                        />
                    </View>
                    <View style={Styles.buttonStyle}>
                        <Button
                            color={Res.colors.gray}
                            onPress={() => {
                                this.onSaveButton();
                            }}
                            title={I18n.t("signature_save")}
                            disabled={this.state.saveDisable}
                        />
                    </View>
                </View>
                <View style={Styles.modalSignature}>
                    <SignatureCapture
                        style={Styles.signature}
                        ref={this.setSignatureRef}
                        onSaveEvent={this.onSaveEvent}
                        onDragEvent={this.onDragEvent}
                        saveImageFileInExtStorage
                        showNativeButtons={false}
                        showTitleLabel
                        viewMode="landscape"// "landscape/portrait"
                        minStrokeWidth={4}
                        maxStrokeWidth={8}
                    />
                </View>
            </View>
        );
    }
}

Signature.propTypes = {
    sendSignature: PropTypes.func.isRequired,
    changeSignature: PropTypes.func
};

Signature.defaultProps = {
    changeSignature: () => {}
};

export default Signature;
