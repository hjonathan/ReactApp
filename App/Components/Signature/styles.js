import {StyleSheet} from "react-native";
import * as Res from "../../Assets/resources";

export default StyleSheet.create({
    modalContainer: {
        flex: 1,
        flexDirection: "column",
        borderRadius: 5,
        backgroundColor: "white"
    },
    modalOptions: {
        padding: 4,
        borderColor: Res.colors.gray,
        borderBottomWidth: 2,
        flexDirection: "row"
    },
    buttonStyle: {
        margin: 4
    },
    modalSignature: {
        flex: 1,
        flexDirection: "row"
    },
    signature: {
        borderColor: Res.colors.lightColor,
        borderWidth: 1,
        flex: 1,
        width: null,
        height: null
    }
});
