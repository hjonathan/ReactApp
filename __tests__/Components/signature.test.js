import 'react-native';
import React from 'react';
import {Button} from "react-native";
import toJson from 'enzyme-to-json';
import {shallow} from 'enzyme';
import Signature from '../../App/Components/Signature';

jest.mock('react-native-orientation', () => require.requireActual('../../__mocks__/react-native-orientation').default);

describe('Component >>> Signature', () => {
    const changeSignature = () => {};
    const sendSignature = () => {};
    let signatureView;
    beforeEach(() => {
        signatureView = shallow(<Signature changeSignature={changeSignature} sendSignature={sendSignature}/>);
    });

    test('Render Signature', () => {
        expect(toJson(signatureView)).toMatchSnapshot();
    });

    test("Props coverage componentWillUnmount", () => {
        expect(signatureView.instance().componentWillUnmount()).toBeUndefined();
    });

    test('Signature test onCancelButton', () => {
        signatureView.instance().setState({orientation: "LANDSCAPE"});
        expect(toJson(signatureView.instance().onCancelButton())).toBeUndefined();
        signatureView.instance().setState({orientation: "PORTRAIT"});
        expect(toJson(signatureView.instance().onCancelButton())).toBeUndefined();
    });

    test('Signature test onResetButton', () => {
        signatureView.instance().signature = {
            resetImage: () => {}
        };
        expect(toJson(signatureView.instance().onResetButton())).toBeUndefined();
    });

    test('Signature test onSaveButton', () => {
        signatureView.instance().signature = {
            saveImage: () => {}
        };
        signatureView.instance().setState({orientation: "LANDSCAPE"});
        expect(toJson(signatureView.instance().onSaveButton())).toBeUndefined();
        signatureView.instance().setState({orientation: "PORTRAIT"});
        expect(toJson(signatureView.instance().onSaveButton())).toBeUndefined();
    });

    test('Signature test onSaveEvent', () => {
        expect(toJson(signatureView.instance().onSaveEvent())).toBeUndefined();
    });

    test('Signature test onDragEvent', () => {
        expect(toJson(signatureView.instance().onDragEvent())).toBeUndefined();
    });

    test('Signature test setSignatureRef', () => {
        expect(toJson(signatureView.instance().setSignatureRef({}))).toBeUndefined();
    });

    test("Signature test render", () => {
        const viewJson = toJson(signatureView.instance().render());
        expect(viewJson).toMatchSnapshot();
        signatureView.instance().signature = {
            resetImage: () => {},
            saveImage: () => {}
        };
        expect(signatureView.find(Button).length).toEqual(3);
        signatureView.find(Button).at(0).props().onPress();
        signatureView.find(Button).at(1).props().onPress();
        signatureView.find(Button).at(2).props().onPress();
    })
});
