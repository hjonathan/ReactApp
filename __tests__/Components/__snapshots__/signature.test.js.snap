// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component >>> Signature Render Signature 1`] = `
<Component
  style={
    Object {
      "backgroundColor": "white",
      "borderRadius": 5,
      "flex": 1,
      "flexDirection": "column",
    }
  }
>
  <Component
    style={
      Object {
        "borderBottomWidth": 2,
        "borderColor": "#BABABA",
        "flexDirection": "row",
        "padding": 4,
      }
    }
  >
    <Component
      style={
        Object {
          "margin": 4,
        }
      }
    >
      <Button
        color="#BABABA"
        onPress={[Function]}
        title="signature_cancel"
      />
    </Component>
    <Component
      style={
        Object {
          "margin": 4,
        }
      }
    >
      <Button
        color="#BABABA"
        onPress={[Function]}
        title="signature_clear"
      />
    </Component>
    <Component
      style={
        Object {
          "margin": 4,
        }
      }
    >
      <Button
        color="#BABABA"
        disabled={true}
        onPress={[Function]}
        title="signature_save"
      />
    </Component>
  </Component>
  <Component
    style={
      Object {
        "flex": 1,
        "flexDirection": "row",
      }
    }
  >
    <SignatureCapture
      maxStrokeWidth={8}
      minStrokeWidth={4}
      onDragEvent={[Function]}
      onSaveEvent={[Function]}
      saveImageFileInExtStorage={true}
      showNativeButtons={false}
      showTitleLabel={true}
      style={
        Object {
          "borderColor": "#3397E1",
          "borderWidth": 1,
          "flex": 1,
          "height": null,
          "width": null,
        }
      }
      viewMode="landscape"
    />
  </Component>
</Component>
`;

exports[`Component >>> Signature Signature test render 1`] = `undefined`;
